### PR TITLE
Added torch.linalg.tensorsolve

### PR DESCRIPTION
**Pull request by IvanYashchuk**
This PR adds `torch.linalg.tensorsolve` function that matches `numpy.linalg.tensorsolve`.

Ref #42666.

### CODE REVIEW

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -8319,6 +8319,10 @@
   python_module: linalg
   variants: function
 
+- func: linalg_tensorsolve(Tensor self, Tensor input2, int[]? axes=None) -> Tensor
```
Comment (mruberry) [inline aten/src/ATen/native/native_functions.yaml:8322]: "input2" -> "other" to be consistent with other PyTorch functions.

"axes" -> "dims" to be consistent with other PyTorch functions. Note that PyTorch's bindings should remap "axes->dims", too.

File: test/test_linalg.py [hunk 2]
```diff
@@ -921,6 +922,39 @@ def test_nuclear_norm_exceptions_old(self, device):
         self.assertRaisesRegex(RuntimeError, "duplicate or invalid", torch.norm, x, "nuc", (0, 0))
         self.assertRaisesRegex(IndexError, "Dimension out of range", torch.norm, x, "nuc", (0, 2))
 
+    @skipCUDAIfNoMagma
+    @skipCPUIfNoLapack
+    @unittest.skipIf(not TEST_NUMPY, "Numpy not found")
```
Comment (mruberry) [inline test/test_linalg.py:927]: Our test suite now requires NumPy so we don't need these skips anymore

File: test/test_linalg.py [hunk 3]
```diff
@@ -921,6 +922,39 @@ def test_nuclear_norm_exceptions_old(self, device):
...
+    @dtypes(torch.float, torch.double, torch.cfloat, torch.cdouble)
+    @dtypesIfCPU(torch.float, torch.double, torch.cfloat, torch.cdouble)
```
Comment (mruberry) [inline test/test_linalg.py:929]: This doesn't need a CPU dtype override since the dtypes are the same as the default dtypes (above)

File: test/test_linalg.py [hunk 4]
```diff
@@ -921,6 +922,39 @@ def test_nuclear_norm_exceptions_old(self, device):
...
+            self.assertEqual(result, expected)
+
+        a_shapes = [(2, 3, 6), (3, 4, 4, 3)]
+        axes = [None, (0, 2)]
+        for a_shape, ax in itertools.product(a_shapes, axes):
+            run_test(a_shape, ax)
+
+    # TODO: once "solve_cuda" supports complex dtypes, they shall be added to above test
+    @unittest.expectedFailure
+    @unittest.skipIf(not TEST_NUMPY, "NumPy not found")
```
Comment (mruberry) [inline test/test_linalg.py:947]: Can remove this

File: test/test_linalg.py [hunk 5]
```diff
@@ -921,6 +922,39 @@ def test_nuclear_norm_exceptions_old(self, device):
...
+        def run_test(a_shape, axes):
+            a = torch.rand(a_shape, dtype=dtype, device=device)
+            b = torch.rand(a_shape[:2], dtype=dtype, device=device)
+            result = torch.linalg.tensorsolve(a, b, axes=axes)
+            expected = np.linalg.tensorsolve(a.cpu().numpy(), b.cpu().numpy(), axes=axes)
+            self.assertEqual(result, expected)
+
+        a_shapes = [(2, 3, 6), (3, 4, 4, 3)]
+        axes = [None, (0, 2)]
+        for a_shape, ax in itertools.product(a_shapes, axes):
```
Comment (mruberry) [inline test/test_linalg.py:942]: Nice enumeration over the product space here.

File: test/test_linalg.py [hunk 6]
```diff
@@ -921,6 +922,39 @@ def test_nuclear_norm_exceptions_old(self, device):
...
+    @unittest.skipIf(not TEST_NUMPY, "Numpy not found")
+    @dtypes(torch.float, torch.double, torch.cfloat, torch.cdouble)
+    @dtypesIfCPU(torch.float, torch.double, torch.cfloat, torch.cdouble)
+    @dtypesIfCUDA(torch.float, torch.double)
+    @precisionOverride({torch.float: 1e-4, torch.cfloat: 1e-4})
+    def test_tensorsolve(self, device, dtype):
+        def run_test(a_shape, axes):
+            a = torch.rand(a_shape, dtype=dtype, device=device)
+            b = torch.rand(a_shape[:2], dtype=dtype, device=device)
+            result = torch.linalg.tensorsolve(a, b, axes=axes)
```
Comment (mruberry) [inline test/test_linalg.py:936]: Update to "dims=" (see above note)

File: torch/csrc/api/include/torch/linalg.h [hunk 7]
```diff
@@ -53,4 +57,18 @@ inline Tensor& linalg_norm_out(Tensor& result, const Tensor& self, std::string o
   return detail::norm_out(result, self, ord, opt_dim, keepdim, opt_dtype);
 }
 
+/// Solving of the tensor equation `a x = b`
```
Comment (mruberry) [inline torch/csrc/api/include/torch/linalg.h:60]: "Solving of the tensor equation" -> "Solves the equation"

File: torch/linalg/__init__.py [hunk 8]
```diff
@@ -140,3 +140,33 @@
     >>> LA.norm(m[0, :, :]), LA.norm(m[1, :, :])
     (tensor(3.7417), tensor(11.2250))
 """)
+
+tensorsolve = _add_docstr(_linalg.linalg_tensorsolve, r"""
+linalg.tensorsolve(a, b, axes=None) -> Tensor
```
Comment (mruberry) [inline torch/linalg/__init__.py:145]: tensorsolve(input, other, dims=None)

File: torch/linalg/__init__.py [hunk 9]
```diff
@@ -140,3 +140,33 @@
...
+
+Computes the solution to the tensor equation ``a x = b`` for ``x``.
```
Comment (mruberry) [inline torch/linalg/__init__.py:147]: "Computes a tensor \`x\` such that \`\`input @ x = other\`\`."

I think that will render correctly?

File: torch/linalg/__init__.py [hunk 10]
```diff
@@ -140,3 +140,33 @@
...
+
+Suppose that `:attr:`b` is equivalent to ``torch.tensordot(a, x)``.
```
Comment (mruberry) [inline torch/linalg/__init__.py:149]: If the initial sentence is updated I don't think this sentence is needed?

File: torch/linalg/__init__.py [hunk 11]
```diff
@@ -140,3 +140,33 @@
...
+This function computes tensor ``x`` from `:attr:`a` and `:attr:`b`.
+
+Args:
+    a (Tensor): left-hand-side tensor of size ``b.shape + Q``.
+        ``Q`` is the shape of that sub-tensor of `:attr:`a` consisting of the appropriate
+        number of its rightmost indices, and must be such that ``prod(Q) == prod(b.shape)``.
+    b (Tensor): right-hand-side tensor of any size.
```
Comment (mruberry) [inline torch/linalg/__init__.py:156]: "solution to tensordot(:attr:\`input\`, x). May be any size."

Is that true that it may actually be any size? Will an empty tensor work, for example?

File: torch/linalg/__init__.py [hunk 12]
```diff
@@ -140,3 +140,33 @@
...
+tensorsolve = _add_docstr(_linalg.linalg_tensorsolve, r"""
+linalg.tensorsolve(a, b, axes=None) -> Tensor
+
+Computes the solution to the tensor equation ``a x = b`` for ``x``.
+
+Suppose that `:attr:`b` is equivalent to ``torch.tensordot(a, x)``.
+This function computes tensor ``x`` from `:attr:`a` and `:attr:`b`.
+
+Args:
+    a (Tensor): left-hand-side tensor of size ``b.shape + Q``.
```
Comment (mruberry) [inline torch/linalg/__init__.py:153]: I'm not sure what this is trying to say. I like that it's trying to describe the input shape but I'm not following the description.

File: torch/linalg/__init__.py [hunk 13]
```diff
@@ -140,3 +140,33 @@
...
+        ``Q`` is the shape of that sub-tensor of `:attr:`a` consisting of the appropriate
+        number of its rightmost indices, and must be such that ``prod(Q) == prod(b.shape)``.
+    b (Tensor): right-hand-side tensor of any size.
+    axes (Tuple[int]): axes (dimensions) of `:attr:`a` to reorder to the right, before solve.
```
Comment (mruberry) [inline torch/linalg/__init__.py:157]: The use of the dims argument will have to be mentioned in the description of the function above, too. "reorder" is an imprecise term that will need careful clarification.

File: torch/linalg/__init__.py [hunk 14]
```diff
@@ -140,3 +140,33 @@
...
+        If None (default), no reordering is done.
+
+    Returns:
```
Comment (mruberry) [inline torch/linalg/__init__.py:160]: I don't think this needs a returns section.

File: torch/linalg/__init__.py [hunk 15]
```diff
@@ -140,3 +140,33 @@
...
+        (Tensor): The tensor with shape ``Q`` such that ``b.shape + Q == a.shape``.
+
+Example::
+
+    >>> a = torch.eye(2 * 3 * 4).reshape((2 * 3, 4, 2, 3, 4))
+    >>> b = torch.randn(2 * 3, 4)
+    >>> x = torch.linalg.tensorsolve(a, b)
+    >>> x.shape
+    torch.Size([2, 3, 4])
+    >>> torch.allclose(torch.tensordot(a, x, dims=3), b)
```
Comment (mruberry) [inline torch/linalg/__init__.py:170]: It's important to relate the dims argument here to the dims argument to this function.

File: test/test_linalg.py [hunk 16]
```diff
@@ -921,6 +922,39 @@ def test_nuclear_norm_exceptions_old(self, device):
         self.assertRaisesRegex(RuntimeError, "duplicate or invalid", torch.norm, x, "nuc", (0, 0))
         self.assertRaisesRegex(IndexError, "Dimension out of range", torch.norm, x, "nuc", (0, 2))
 
+    @skipCUDAIfNoMagma
```
Comment (mruberry) [inline test/test_linalg.py:925]: Update common_methods_invocations.py's method_tests, too.

File: torch/linalg/__init__.py [hunk 17]
```diff
@@ -140,3 +140,33 @@
...
 """)
+
+tensorsolve = _add_docstr(_linalg.linalg_tensorsolve, r"""
+linalg.tensorsolve(a, b, axes=None) -> Tensor
+
+Computes the solution to the tensor equation ``a x = b`` for ``x``.
+
+Suppose that `:attr:`b` is equivalent to ``torch.tensordot(a, x)``.
+This function computes tensor ``x`` from `:attr:`a` and `:attr:`b`.
+
```
Comment (mruberry) [inline torch/linalg/__init__.py:151]: Add a brief "Supports..." statement here.

File: aten/src/ATen/native/native_functions.yaml (hunk 1 above)
Comment (mruberry) [reply aten/src/ATen/native/native_functions.yaml:8322]: What's the plan for out support?

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 18]
```diff
@@ -1601,6 +1601,33 @@ Tensor& linalg_norm_out(Tensor& result, const Tensor& self, std::string ord, opt
   return linalg_norm_out_impl(result, self, c10::nullopt, ord, opt_dim, keepdim, opt_dtype);
 }
 
+Tensor linalg_tensorsolve(const Tensor& self, const Tensor& input2, optional<IntArrayRef> axes) {
```
Comment (mruberry) [inline aten/src/ATen/native/LinearAlgebra.cpp:1604]: Add a brief comment describing the approach

File: test/test_linalg.py [hunk 19]
```diff
@@ -921,6 +922,39 @@ def test_nuclear_norm_exceptions_old(self, device):
...
+    @precisionOverride({torch.float: 1e-4, torch.cfloat: 1e-4})
+    def test_tensorsolve(self, device, dtype):
+        def run_test(a_shape, axes):
+            a = torch.rand(a_shape, dtype=dtype, device=device)
+            b = torch.rand(a_shape[:2], dtype=dtype, device=device)
+            result = torch.linalg.tensorsolve(a, b, axes=axes)
+            expected = np.linalg.tensorsolve(a.cpu().numpy(), b.cpu().numpy(), axes=axes)
+            self.assertEqual(result, expected)
+
+        a_shapes = [(2, 3, 6), (3, 4, 4, 3)]
```
Comment (mruberry) [inline test/test_linalg.py:940]: Are there odd/extremal edge cases to test, too, like how empty tensors (on both the lhs and rhs) are handled?

**Review by mruberry — CHANGES_REQUESTED** [review]
Hey @IvanYashchuk!

This is a good PR that's already near completion. It needs some consistency updates (arg names, out support), maybe a few more test cases, and some additional doc clarity. As I'm sure @kurtamohler or @peterbell10 will tell you I'm a stickler for super-precise clear documentation. Writing documentation is hard and the current docs aren't bad, but I think it's important we spend the time to make our documentation great. It is, after all, how many PyTorch users learn how to use the framework.

File: torch/linalg/__init__.py (hunk 12 above)
Comment (IvanYashchuk) [reply torch/linalg/__init__.py:153]: Maybe the confusion comes from using the word "size", I don't know what is the preferred term now, but I was referring to the shape of the tensor here (`tensor.size() == tensor.shape`).
`other.shape` is a tuple in NumPy and `torch.Size` in PyTorch and `Q` is a tuple, then `input.shape == other.shape + Q`.
I should change `Q`->`x.shape`.

File: torch/linalg/__init__.py (hunk 11 above)
Comment (IvanYashchuk) [reply torch/linalg/__init__.py:156]: Size == shape here.

It's true that can be any size. Choosing `b` (or `other`) puts restrictions on shape of `a` (or `input`).
An empty tensor would work if `lda` parameter (4th argument) in `lapackSolve` call was handled correctly.

**Comment by IvanYashchuk** [comment]
In order to handle empty input cases correctly I fixed an error related to `lda` parameter in `apply_solve` in `BatchLinearAlgebra.{cpp, cu}`

File: torch/linalg/__init__.py (hunk 15 above)
Comment (IvanYashchuk) [reply torch/linalg/__init__.py]: They are not related. `dims` argument to the current function tells how to reorder the dimensions of input.
I've changed the documentation to indicate that `dims` argument to `tensordot` should be `len(x.shape)` or `x.ndim`.

File: aten/src/ATen/native/native_functions.yaml (hunk 1 above)
Comment (IvanYashchuk) [reply aten/src/ATen/native/native_functions.yaml]: I wasn't thinking about out support, since NumPy doesn't do that. I guess it can be added to this PR.

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 20]
```diff
@@ -1601,6 +1601,33 @@ Tensor& linalg_norm_out(Tensor& result, const Tensor& self, std::string ord, opt
...
+
+  // prepare the permutation vector with given axes and permute self_
+  if (axes.has_value()) {
+    std::vector<int64_t> permutation_axes(ndim);
+    std::iota(permutation_axes.begin(), permutation_axes.end(), 0);
+    for (int64_t k : axes.value().vec()) {
+      auto it = std::find(permutation_axes.begin(), permutation_axes.end(), k);
+      std::rotate(it, it + 1, permutation_axes.end());
+    }
+    self_ = self_.permute(permutation_axes);
```
[the remark this answers is not in the document]
Comment (IvanYashchuk) [reply aten/src/ATen/native/LinearAlgebra.cpp]: Thank you! I was not aware of this function.

File: test/test_linalg.py [hunk 21]
```diff
@@ -921,6 +922,69 @@ def test_nuclear_norm_exceptions_old(self, device):
...
+        for a_shape, d in itertools.product(a_shapes, dims):
+            run_test(a_shape, d)
+
+    @skipCUDAIfNoMagma
+    @skipCPUIfNoLapack
+    @dtypes(torch.float, torch.double, torch.cfloat, torch.cdouble)
+    @dtypesIfCUDA(torch.float, torch.double)
+    def test_tensorsolve_empty(self, device, dtype):
+        # Check for empty inputs. NumPy does not work for these cases.
+        a = torch.empty(0, 0, 1, 2, 3, 0)
```
Comment (IvanYashchuk) [inline test/test_linalg.py:955]:
```suggestion
        a = torch.empty(0, 0, 1, 2, 3, 0, dtype=dtype, device=device)
```

File: torch/csrc/api/include/torch/linalg.h [hunk 22]
```diff
@@ -28,6 +28,10 @@ inline Tensor& norm_out(Tensor& result, const Tensor& self, std::string ord, opt
   return torch::linalg_norm_out(result, self, ord, opt_dim, keepdim, opt_dtype);
 }
 
+inline Tensor tensorsolve(const Tensor& self, const Tensor& other, optional<IntArrayRef> dims) {
```
Comment (IvanYashchuk) [inline torch/csrc/api/include/torch/linalg.h:31]: - [ ] Add also `tensorsolve_out` here.

**Comment by IvanYashchuk** [comment]
@mruberry, I've added a few changes to this PR. I think it's in better shape now.
I updated the documentation, I hope now it's more clear even though that's a complicated function.
`out=` support was added as well. I updated the tests to include empty, non-contiguous inputs and error checks for invalid input.

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 23]
```diff
@@ -1602,6 +1602,69 @@ Tensor& linalg_norm_out(Tensor& result, const Tensor& self, std::string ord, opt
...
+
+  // result_shape is self_.sizes[-(an-other.dim):]
+  std::vector<int64_t> result_shape = self_.sizes().slice(other.dim(), ndim - other.dim()).vec();
+
+  int64_t result_product = std::accumulate(result_shape.begin(), result_shape.end(), int64_t{1}, std::multiplies<int64_t>());
+  int64_t other_product = std::accumulate(other.sizes().begin(), other.sizes().end(), int64_t{1}, std::multiplies<int64_t>());
+
+  // Check whether the self tensor can be reshaped to the 2D square matrix
+  TORCH_CHECK(result_product == other_product,
+    "Expected self to be a 'square' tensor such that prod(result.shape) == prod(other.shape), but got ",
```
Comment (mruberry) [inline aten/src/ATen/native/LinearAlgebra.cpp:1633]: This is a little wonky since "result" is not user-visible and it would be more accurate to say, as the comment does, that self can be reshaped into a `square` tensor for purposes of this problem, not that it is square. I think taking the time to add more detail here and relate the requirement back to self and other would be helpful.

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 24]
```diff
@@ -1602,6 +1602,69 @@ Tensor& linalg_norm_out(Tensor& result, const Tensor& self, std::string ord, opt
...
+
+  Tensor self_ = self;
+  // move dimensions of `self_` from `dims` to the end
+  if (dims.has_value()) {
+    DimVector dest_axes(dims.value().size());
+    std::iota(dest_axes.begin(), dest_axes.end(), self.dim() - dest_axes.size());
+    self_ = at::movedim(self_, dims.value(), dest_axes);
+  }
+
+  auto expected_result_shape = self_.sizes().slice(other.dim(), self.dim() - other.dim());
```
Comment (mruberry) [inline aten/src/ATen/native/LinearAlgebra.cpp:1658]: This is OK. This check could also occur after the computation by verifying that result_tmp and result have the same shape.

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 25]
```diff
@@ -1602,6 +1602,69 @@ Tensor& linalg_norm_out(Tensor& result, const Tensor& self, std::string ord, opt
...
+  TORCH_CHECK(result.sizes().equals(expected_result_shape),
+    "Expected result tensor to have size of ", expected_result_shape, ", but got tensor of size ", result.sizes());
+
+  // We've already reordered self with dims, let's pass that result
+  Tensor result_tmp = at::linalg_tensorsolve(self_, other, c10::nullopt);
+  result.copy_(result_tmp);
```
Comment (mruberry) [inline aten/src/ATen/native/LinearAlgebra.cpp:1664]: An interesting side affect of this approach is that it will allow copying to a result tensor with a different dtype than result_tmp. This pattern of computing a result and then copying to an out tensor is common enough we should consider making a helper function for it.

File: torch/linalg/__init__.py [hunk 26]
```diff
@@ -140,3 +140,51 @@
     >>> LA.norm(m[0, :, :]), LA.norm(m[1, :, :])
     (tensor(3.7417), tensor(11.2250))
 """)
+
+tensorsolve = _add_docstr(_linalg.linalg_tensorsolve, r"""
+linalg.tensorsolve(input, other, dims=None, *, out=None) -> Tensor
+
+Computes a tensor ``x`` such that ``tensordot(input, x, dims=x.ndim) = other``.
+The resulting tensor ``x`` has the shape equal to ``Q``, where ``Q`` is the shape
```
Comment (mruberry) [inline torch/linalg/__init__.py:148]: "has the shape equal to Q" -> "has the same shape as Q"

File: torch/linalg/__init__.py [hunk 27]
```diff
@@ -140,3 +140,51 @@
     >>> LA.norm(m[0, :, :]), LA.norm(m[1, :, :])
     (tensor(3.7417), tensor(11.2250))
 """)
+
+tensorsolve = _add_docstr(_linalg.linalg_tensorsolve, r"""
+linalg.tensorsolve(input, other, dims=None, *, out=None) -> Tensor
+
+Computes a tensor ``x`` such that ``tensordot(input, x, dims=x.ndim) = other``.
```
Comment (mruberry) [inline torch/linalg/__init__.py:147]: Great first sentence.

File: torch/linalg/__init__.py (hunk 26 above)
Comment (mruberry) [reply torch/linalg/__init__.py:148]: Before describing the shape of a result, I think this should start a new paragraph and discuss the shape requirement for input and other. That should make defining the shape of x easier. 

As I understand it, if the shape of input is (a, b, c, d, ... z), then the shape of other must (a, b, c, d...) and the shape of x will be (..., x, y z). Is that correct?

File: torch/linalg/__init__.py [hunk 28]
```diff
@@ -140,3 +140,51 @@
...
+The resulting tensor ``x`` has the shape equal to ``Q``, where ``Q`` is the shape
+of the sub-tensor of `:attr:`input` consisting of the appropriate number of
+its rightmost dimensions such that ``input.shape = other.shape + Q``.
+Before computing ``x``, dimensions of `:attr:`input` can be moved to the right
+using :attr:`dims` to match the shape requirements.
+
+Supports real and, only on the CPU, complex inputs.
```
Comment (mruberry) [inline torch/linalg/__init__.py:154]: "Supports real" -> "Supports float"

File: torch/linalg/__init__.py [hunk 29]
```diff
@@ -140,3 +140,51 @@
...
+
+.. note:: If :attr:`input` is not a 'square' tensor, meaning it does not satisfy the requirement
```
Comment (mruberry) [inline torch/linalg/__init__.py:156]: This note doesn't seem correct since it doesn't consider the dims argument, for example. Also, I think we should avoid describing the input as "square," since square suggests that input is a nxn matrix.

File: torch/linalg/__init__.py [hunk 30]
```diff
@@ -140,3 +140,51 @@
...
+          ``prod(x.shape) == prod(other.shape)``, then a RuntimeError will be thrown.
+
+Args:
+    input (Tensor): left-hand-side tensor of shape ``other.shape + Q``.
+        ``Q`` is the shape of that sub-tensor of `:attr:`input` consisting of the appropriate
+        number of its rightmost dimensions, and must be such that ``prod(Q) == prod(other.shape)``.
+    other (Tensor): right-hand-side tensor of any shape.
```
Comment (mruberry) [inline torch/linalg/__init__.py:163]: Isn't this constrained just as input is above?

File: torch/linalg/__init__.py [hunk 31]
```diff
@@ -140,3 +140,51 @@
...
+    dims (Tuple[int]): dimensions of `:attr:`input` to be moved to the right, before solve.
```
Comment (mruberry) [inline torch/linalg/__init__.py:164]: Saying the dimensions are "moved to right" is a little vague (how far to the right? in what order? what happens to the other dims?). Can this describe the behavior in terms of another function, like movedim? Is it, for example, that each dim is "movedim'd" to the -1 position in the order it appears it the tuple?

**Comment by mruberry** [comment]
Hey Ivan, 

Overall this PR and its latest updates look great. This is a tricky function to describe, however. I made a few more comments/suggestions about error messages and the docs.

File: aten/src/ATen/native/LinearAlgebra.cpp (hunk 24 above)
Comment (IvanYashchuk) [reply aten/src/ATen/native/LinearAlgebra.cpp:1658]: For large inputs the computation can take a long time. I thought it's a good idea to exit the function before doing actual computations if those will be wasteful.

File: aten/src/ATen/native/LinearAlgebra.cpp (hunk 25 above)
Comment (IvanYashchuk) [reply aten/src/ATen/native/LinearAlgebra.cpp:1664]: I don't know what's the expected behavior in general, so there is a check now that doesn't allow different dtypes.
https://github.com/pytorch/pytorch/pull/46142/files/497bc2241ea7ceae548277aac989849ad5a1bb0b#diff-db2b91c51cffd1fba933933ac20de256f5f0ced095890a1a49a17d73037001a2R1648

File: torch/linalg/__init__.py (hunk 28 above)
Comment (IvanYashchuk) [reply torch/linalg/__init__.py:154]: Both real and complex numbers are floating-point numbers. What about 
```suggestion
Supports real-valued and, only on the CPU, complex-valued inputs.
```

File: torch/linalg/__init__.py (hunk 26 above)
Comment (IvanYashchuk) [reply torch/linalg/__init__.py:148]: > As I understand it, if the shape of input is (a, b, c, d, ... z), then the shape of other must (a, b, c, d...) and the shape of x will be (..., x, y z). Is that correct?

That is correct, but ...
Here is how I understand the current NumPy doc. First, the `other` tensor is fixed, that's why it can be of any shape. Then based on `other.shape` the requirements on `input` are posed. If `other.shape` is `(a, b, c, d)` then:
- `input.shape` must start with `a, b, c, d`
  - at this stage `torch.movedim(input, dims, (-(len(dims)+1), ..., -2, -1))` can be used to match the requirement.
 (moving dimensions specified in `dims` to the right end keeping the order)
- `input.shape = (a, b, c, d) + (..., x, y, z) = (a, b, c, d, ..., x, y, z)`
- `x.shape = (..., x, y, z)`

For the function to work `input` must satisfy the requirement `prod(..., x, y, z)==prod(a, b, c, d)`.

Alternatively, we can start with `input` and `other.ndim`, then:
- `other.shape` must be such that `other.shape == input.shape[0:other.ndim]` (that point is a bit problematic since we pose the requirement on `other` using `other.ndim`.
- `x.shape = input.shape[other.ndim:]`

`input` must satisfy the requirement `prod(input.shape[other.ndim:] == input.shape[:other.ndim]`.

I think that the second variant is better, because then we don't introduce any `Q`, and the "squareness" requirement is directly relatable to `tensorinv` function with `ind = other.ndim`.

File: aten/src/ATen/native/LinearAlgebra.cpp (hunk 23 above)
Comment (IvanYashchuk) [reply aten/src/ATen/native/LinearAlgebra.cpp:1633]: I've changed the message to `Expected self to satisfy the requirement prod(self.shape[other.ndim:]) == prod(self.shape[:other.ndim]), but got `

File: torch/linalg/__init__.py (hunk 31 above)
Comment (IvanYashchuk) [reply torch/linalg/__init__.py]: How about this?
```
    dims (Tuple[int]): dimensions of `:attr:`input` to be moved to the end keeping the order,
                       i.e. ``movedim(input, dims, range(len(dims) - input.ndim, 0)), before solve.
                       If None (default), no moving of dimensions is done.
```

File: torch/linalg/__init__.py (hunk 29 above)
Comment (IvanYashchuk) [reply torch/linalg/__init__.py]: I've changed this note to 
```
.. note:: If :attr:`input` does not satisfy the requirement
          ``prod(input.shape[other.ndim:]) == prod(input.shape[:other.ndim])``
          after (optionally) moving the dimensions using using :attr:`dims`, then a RuntimeError will be thrown.
```

**Comment by IvanYashchuk** [comment]
So JIT started to fail, need to fix that.

**Comment by IvanYashchuk** [comment]
@mruberry, this is ready for another round.
I've updated the general description of the function with the second option from here https://github.com/pytorch/pytorch/pull/46142#discussion_r510054412
`dims` argument is described as here https://github.com/pytorch/pytorch/pull/46142#discussion_r510072369

Updated `linalg_tensorsolve_out` to use `at::native::resize_output`.
Explicit `dispatch: Math` is now used `native_functions.yml`.
Added non-contiguous out= tests cases.

File: torch/csrc/api/include/torch/linalg.h [hunk 32]
```diff
@@ -53,4 +61,22 @@ inline Tensor& linalg_norm_out(Tensor& result, const Tensor& self, std::string o
   return detail::norm_out(result, self, ord, opt_dim, keepdim, opt_dtype);
 }
 
+/// Solves the tensor equation `a x = b`
```
Comment (mruberry) [inline torch/csrc/api/include/torch/linalg.h:64]: This line needs an update to reflect the updates in the regular documentation.

File: torch/linalg/__init__.py [hunk 33]
```diff
@@ -140,3 +140,49 @@
     >>> LA.norm(m[0, :, :]), LA.norm(m[1, :, :])
     (tensor(3.7417), tensor(11.2250))
 """)
+
+tensorsolve = _add_docstr(_linalg.linalg_tensorsolve, r"""
+linalg.tensorsolve(input, other, dims=None, *, out=None) -> Tensor
+
+Computes a tensor ``x`` such that ``tensordot(input, x, dims=x.ndim) = other``.
+The resulting tensor ``x`` has the shape equal to ``input[other.ndim:]``.
```
Comment (mruberry) [inline torch/linalg/__init__.py:148]: "the shape equal to" -> "the same shape as"

File: torch/linalg/__init__.py [hunk 34]
```diff
@@ -140,3 +140,49 @@
...
+Before computing ``x``, dimensions of :attr:`input` can be moved using :attr:`dims` to match the shape requirements.
```
Comment (mruberry) [inline torch/linalg/__init__.py:149]: Since the shape requirements haven't been introduced yet and "moved" is not well-defined, I suggest removing this sentence and leaving the documentation of dims to the more detailed note below.

File: torch/linalg/__init__.py [hunk 35]
```diff
@@ -140,3 +140,49 @@
...
+
+Supports real-valued and, only on the CPU, complex-valued inputs.
+
+.. note:: If :attr:`input` does not satisfy the requirement
+          ``prod(input.shape[other.ndim:]) == prod(input.shape[:other.ndim])``
+          after (optionally) moving the dimensions using using :attr:`dims`, then a RuntimeError will be thrown.
```
Comment (mruberry) [inline torch/linalg/__init__.py:155]: "using using :attr:\`dims\`" -> "using :attr:\`dims\` (see below for details)"

File: torch/linalg/__init__.py [hunk 36]
```diff
@@ -140,3 +140,49 @@
...
+
+Args:
+    input (Tensor): left-hand-side tensor, it must satisfy the requirement
```
Comment (mruberry) [inline torch/linalg/__init__.py:158]: Put quotation marks around "left-hand-side" so it's clear that's an informal term.

File: torch/linalg/__init__.py [hunk 37]
```diff
@@ -140,3 +140,49 @@
...
+                    ``prod(input.shape[other.ndim:]) == prod(input.shape[:other.ndim])``.
+    other (Tensor): right-hand-side tensor of shape ``input.shape[other.ndim]``.
```
Comment (mruberry) [inline torch/linalg/__init__.py:160]: Quotes around "right-hand-side" here, too

File: torch/linalg/__init__.py [hunk 38]
```diff
@@ -140,3 +140,49 @@
...
+    dims (Tuple[int]): dimensions of :attr:`input` to be moved to the end keeping the order,
```
Comment (mruberry) [inline torch/linalg/__init__.py:161]: "to be moved to the end keeping the order, ... " -> "to be moved before the computation. Equivalent to calling \`\`input = movedim(input, dims, range(len(dims) - input.ndim, 0))\`\`. If None (default), no dimensions are moved."

**Review by mruberry — APPROVED** [review]
Great work, @IvanYashchuk. I just made a few comments/suggestions that should be easy to implement. Let me know when this is ready to merge!

**Comment by IvanYashchuk** [comment]
Thank you for reviewing this pull request and for all the suggestions, @mruberry!
Now this is ready to merge 🎉

**Comment by mruberry** [comment]
Hey @IvanYashchuk, would you rebase this, please? The internal build tools are hitting an issue that I expect a rebase to solve.

File: torch/testing/_internal/common_methods_invocations.py [hunk 39]
```diff
@@ -1444,6 +1444,8 @@ def method_tests():
         ('__getitem__', torch.randn(S, S, S), (dont_convert([[0, 2, 3], [1, 3, 3],
                                                              torch.LongTensor([0, 0, 2])]),), 'adv_index_var'),
         ('to_sparse', (S, S), (), '', (), (), [], lambda x: x.to_dense()),
+        ('linalg.tensorsolve', (2, S, M), ((2, S),),
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:1447]: Most of our internal build issues have been resolved but there's still an internal problem resolving the references to the tests generated here. Would you just revert the changes to common_methos_invocations.py for a moment? We should move these tests to use OpInfos soon, anyway. The temporary gap in test coverage is OK. 

Sorry to keep bringing these internal tooling issues to you, @IvanYashchuk.

Comment (IvanYashchuk) [reply torch/testing/_internal/common_methods_invocations.py]: Okay, I removed it.
